webui/server/server.py

```diff
@@ -285,85 +268,178 @@ def Write(self):
+
+
+async def get_ws(request):
+  ws = web.WebSocketResponse()
+  await ws.prepare(request)
+
+  print('Client connected')
+  profile_handler.MaybeLoad()
+
+  # The above does emit if there are differences, so have an extra for the
+  # case there are no differences.
+  await ws.send_json([
+    'thresholds',
+    {'thresholds': profile_handler.GetCurThresholds()},
+  ])
+
+  global read_thread, write_thread
+
+  if not read_thread or not read_thread.is_alive():
+    print('Starting Read thread')
+    read_thread = threading.Thread(target=serial_handler.Read)
+    read_thread.daemon = True
+    read_thread.start()
+
+  if not write_thread or not write_thread.is_alive():
+    print('Starting Write thread')
+    write_thread = threading.Thread(target=serial_handler.Write)
+    write_thread.daemon = True
+    write_thread.start()
+
+  # Potentially fetch any threshold values from the microcontroller that
+  # may be out of sync with our profiles.
+  serial_handler.write_queue.put('t\n', block=False)
+
+  queue = asyncio.Queue(maxsize=100)
+  with out_queues_lock:
+    out_queues.add(queue)
+
+  try:
+    queue_task = asyncio.create_task(queue.get())
+    receive_task = asyncio.create_task(ws.receive())
+    connected = True
+
+    while connected:
+      done, pending = await asyncio.wait([
+        queue_task,
+        receive_task,
+      ], return_when=asyncio.FIRST_COMPLETED)
+
+      for task in done:
+        if task == queue_task:
+          msg = await queue_task
+          await ws.send_json(msg)
+
+          queue_task = asyncio.create_task(queue.get())
+        elif task == receive_task:
+          msg = await receive_task
+
+          if msg.type == WSMsgType.TEXT:
+            data = msg.json()
+            action = data[0]
+
+            if action == 'update_threshold':
+              values, index = data[1:]
+              update_threshold(values, index)
+            elif action == 'add_profile':
+              profile_name, thresholds = data[1:]
+              add_profile(profile_name, thresholds)
+            elif action == 'remove_profile':
+              profile_name, = data[1:]
+              remove_profile(profile_name)
+            elif action == 'change_profile':
+              profile_name, = data[1:]
+              change_profile(profile_name)
+          elif msg.type == WSMsgType.CLOSE:
+            connected = False
+            continue
+
+          receive_task = asyncio.create_task(ws.receive())
+  except ConnectionResetError:
+    pass
+  finally:
+    with out_queues_lock:
+      out_queues.remove(queue)
+
+  queue_task.cancel()
+  receive_task.cancel()
+
+  print('Client disconnected')
+
+
+build_dir = os.path.abspath(
+  os.path.join(os.path.dirname(__file__), '..', 'build')
+)
+
+
+async def get_index(request):
+  return web.FileResponse(os.path.join(build_dir, 'index.html'))
+
+
+app = web.Application()
+app.add_routes([
+  web.get('/defaults', get_defaults),
+  web.get('/ws', get_ws),
+  web.get('/', get_index),
```

Comment:
For previous behavior: '/' shouldn't be added if NO_SERIAL is true (because of the node server separation).

Reply:
although I'm sure the difference between .get() and the .static() call below

Reply:
`web.get('/', get_index)` - return build/index.html when `/` is requested
`web.get('/plot', get_index)` - return build/index.html when '/' is requested
`web.static('/', build_dir)` - serve all static files in the build folder: e.g.  requesting '/static/css/2.d9ad5f5c.chunk.css' will serve `build/static/css/2.d9ad5f5c.chunk.css`.